### PR TITLE
refac: CreateBatchEvent uses mediatr

### PR DESCRIPTION
**Pull request by energinet-ajw**
## Description
This PR will be the first to transition from publishing internal domain events using a service bus. To distribute them in-memory using Mediatr.  

- Removed StartCalculationEndpoint 

Upcoming work
- IRequest and INotification are wrapped to IEvent and ICommand
- Next event to refac: BatchCompletedEvent

## References

* #972

### CODE REVIEW

File: source/dotnet/wholesale-api/Application/Batches/StartCalculationJobHandler.cs [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using Energinet.DataHub.Wholesale.Domain.BatchAggregate;
+using Energinet.DataHub.Wholesale.Domain.CalculationDomainService;
+using MediatR;
+
+namespace Energinet.DataHub.Wholesale.Application.Batches;
+
+public class StartCalculationJobHandler : INotificationHandler<BatchCreatedDomainEvent>
```
Comment (BjarkeMeier) [inline source/dotnet/wholesale-api/Application/Batches/StartCalculationJobHandler.cs:21]: Probably another PR, but perhaps:
- Rename domain events to either "BatchCreated" or "BatchCreatedEvent"?
- Introduce an IEventHandler to align terminology in the system?

File: source/dotnet/wholesale-api/Application/Batches/StartCalculationJobHandler.cs [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+    {
+        _calculationDomainService = calculationDomainService;
+    }
+
+    /// <summary>
+    /// Create and start all processes of batches with state <see cref="BatchExecutionState.Submitted"/>.
+    /// </summary>
+    public async Task Handle(BatchCreatedDomainEvent notification, CancellationToken cancellationToken)
+    {
+        await _calculationDomainService.StartAsync(notification.BatchId).ConfigureAwait(false);
```
Comment (BjarkeMeier) [inline source/dotnet/wholesale-api/Application/Batches/StartCalculationJobHandler.cs:36]: What are your thought on whether event handlers are allowed to mutate state (and commit)?

File: source/dotnet/wholesale-api/Application/Batches/StartCalculationJobHandler.cs (hunk 1 above)
Comment (energinet-ajw) [reply source/dotnet/wholesale-api/Application/Batches/StartCalculationJobHandler.cs:21]: Changed to BatchCreatedEvent.
 IEventHandler is in the next PR.

File: source/dotnet/wholesale-api/WebApi.IntegrationTests/WebApi/RequestMatchingHandlerTests.cs [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+{
+    /// <summary>
+    /// Code borrowed from https://matthiaslischka.at/2019/02/25/Testing-MediatR-Registrations/.
+    /// </summary>
+    [Theory]
+    [InlineAutoData(typeof(Root))]
+    [InlineAutoData(typeof(Application.Root))]
+    [InlineAutoData(typeof(Domain.Root))]
+    [InlineAutoData(typeof(Wholesale.Infrastructure.Root))]
+    public void AllRequests_ShouldHaveMatchingHandler(Type type)
```
Comment (BjarkeMeier) [inline source/dotnet/wholesale-api/WebApi.IntegrationTests/WebApi/RequestMatchingHandlerTests.cs:32]:
```suggestion
    public void EachCommand_HasMatchingHandler(Type type)
```

File: source/dotnet/wholesale-api/WebApi.IntegrationTests/WebApi/RequestMatchingHandlerTests.cs [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+    {
+        // Arrange
+        var requestTypes = type.Assembly.GetTypes()
```
Comment (BjarkeMeier) [inline source/dotnet/wholesale-api/WebApi.IntegrationTests/WebApi/RequestMatchingHandlerTests.cs:35]:
```suggestion
        var commandTypes = type.Assembly.GetTypes()
```

File: source/dotnet/wholesale-api/WebApi.IntegrationTests/WebApi/RequestMatchingHandlerTests.cs [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+
+        // Act
+        foreach (var requestType in requestTypes)
+        {
+            // Assert
+            handlerTypes.Should().ContainSingle(handlerType => IsHandlerForRequest(handlerType, requestType), $"Handler for type {requestType} expected");
+        }
+    }
+
+    private static bool IsRequest(Type type)
```
Comment (BjarkeMeier) [inline source/dotnet/wholesale-api/WebApi.IntegrationTests/WebApi/RequestMatchingHandlerTests.cs:51]:
```suggestion
    private static bool IsCommand(Type type)
```

File: source/dotnet/wholesale-api/WebApi.IntegrationTests/WebApi/RequestMatchingHandlerTests.cs [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+    {
+        return typeof(IBaseRequest).IsAssignableFrom(type);
+    }
+
+    private static bool IsIRequestHandler(Type type)
```
Comment (BjarkeMeier) [inline source/dotnet/wholesale-api/WebApi.IntegrationTests/WebApi/RequestMatchingHandlerTests.cs:56]:
```suggestion
    private static bool IsICommandHandler(Type type)
```

File: source/dotnet/wholesale-api/WebApi.UnitTests/Application/Batches/CreateBatchHandlerTests.cs [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+using Xunit.Categories;
+
+namespace Energinet.DataHub.Wholesale.WebApi.UnitTests.Application.Batches;
+
+[UnitTest]
+public class CreateBatchHandlerTests
+{
+    [Theory]
+    [InlineAutoMoqData]
+    public async Task Handle_CallsPublishEvent(
```
Comment (BjarkeMeier) [inline source/dotnet/wholesale-api/WebApi.UnitTests/Application/Batches/CreateBatchHandlerTests.cs:34]:
```suggestion
    public async Task Handle_PublishesBatchCreatedEvent(
```

File: source/dotnet/wholesale-api/WebApi.UnitTests/Application/Batches/CreateBatchHandlerTests.cs [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+        // Act
+        await sut.Handle(batchCommand, default);
+
+        // Assert
+        mediatrMock.Verify(x => x.Publish(It.Is<BatchCreatedDomainEvent>(e => e.BatchId == batch.Id), default));
+    }
+
+    [Theory]
+    [InlineAutoMoqData]
+    public async Task Handle_CallsAddAsync(
```
Comment (BjarkeMeier) [inline source/dotnet/wholesale-api/WebApi.UnitTests/Application/Batches/CreateBatchHandlerTests.cs:54]: What is "Add"? People reading the test name probably don't have a clue without reading the implementation.
```suggestion
    public async Task Handle_AddsBatchToRepository(
```

File: source/dotnet/wholesale-api/WebApi.UnitTests/Application/Batches/Periods.cs [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using NodaTime;
+using NodaTime.Extensions;
+
+namespace Energinet.DataHub.Wholesale.WebApi.UnitTests.Application.Batches;
+
+public static class Periods
```
Comment (BjarkeMeier) [inline source/dotnet/wholesale-api/WebApi.UnitTests/Application/Batches/Periods.cs:20]: Is there any good solution where we can avoid duplicating this code?

File: source/dotnet/wholesale-api/WebApi.UnitTests/Application/Batches/StartCalculationJobHandlerTests.cs [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+using Xunit.Categories;
+
+namespace Energinet.DataHub.Wholesale.WebApi.UnitTests.Application.Batches;
+
+[UnitTest]
+public class StartCalculationJobHandlerTests
+{
+    [Theory]
+    [InlineAutoMoqData]
+    public async Task Handle_ActivatesDomainService(
```
Comment (BjarkeMeier) [inline source/dotnet/wholesale-api/WebApi.UnitTests/Application/Batches/StartCalculationJobHandlerTests.cs:31]: Which domain service?
```suggestion
    public async Task Handle_StartsCalculation(
```

File: source/dotnet/wholesale-api/WebApi/Startup.cs [hunk 11]
```diff
@@ -98,6 +100,9 @@ public void ConfigureServices(IServiceCollection services)
             cfg.RegisterServicesFromAssembly(typeof(Domain.Root).Assembly);
             cfg.RegisterServicesFromAssembly(typeof(Infrastructure.Root).Assembly);
         });
+        services.AddScoped(typeof(IPipelineBehavior<,>), typeof(UnitOfWorkPipelineBehavior<,>));
+        services.AddScoped<IRequestHandler<CreateBatchCommand, Guid>, CreateBatchHandler>();
+        services.AddScoped<INotificationHandler<BatchCreatedDomainEvent>, StartCalculationJobHandler>();
```
Comment (BjarkeMeier) [inline source/dotnet/wholesale-api/WebApi/Startup.cs:105]: How about creating another PR and adding auto-registration using reflection?

File: source/dotnet/wholesale-api/Infrastructure/Persistence/UnitOfWorkPipelineBehavior.cs [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+    where TRequest : IRequest<TResponse>
+{
+    private readonly IUnitOfWork _unitOfWork;
+
+    public UnitOfWorkPipelineBehavior(IUnitOfWork unitOfWork)
+    {
+        _unitOfWork = unitOfWork;
+    }
+
+    public async Task<TResponse> Handle(TRequest request, RequestHandlerDelegate<TResponse> next, CancellationToken cancellationToken)
```
Comment (BjarkeMeier) [inline source/dotnet/wholesale-api/Infrastructure/Persistence/UnitOfWorkPipelineBehavior.cs:30]: Are you 100% confident that if handler A published an event, which is handled by handler B that this pipeline does not commit after B and before A - which may fail afterward? I'd like to see it as a unit test.
Please enlighten me if it's a silly idea/request 😄

File: source/dotnet/wholesale-api/Application/Batches/StartCalculationJobHandler.cs (hunk 2 above)
Comment (BjarkeMeier) [inline source/dotnet/wholesale-api/Application/Batches/StartCalculationJobHandler.cs:36]: So if some handler in the tree of handlers being executed fail then we'll have a situation where the batch wasn't created but the calculation job was started. What are your thoughts on this?

Comment (BjarkeMeier) [reply source/dotnet/wholesale-api/Application/Batches/StartCalculationJobHandler.cs:36]: We talked about this in the team. I recommend that we note that we're aware of this conceptual problem with in-process event handling, but accept the possible consequences in this specific case as they don't seem critical.

File: source/dotnet/wholesale-api/WebApi.UnitTests/Application/Batches/Periods.cs (hunk 9 above)
Comment (energinet-ajw) [reply source/dotnet/wholesale-api/WebApi.UnitTests/Application/Batches/Periods.cs:20]: I take care of this in another PR.

File: source/dotnet/wholesale-api/WebApi/Startup.cs (hunk 11 above)
Comment (energinet-ajw) [reply source/dotnet/wholesale-api/WebApi/Startup.cs]: Correct. They are not needed. They are auto registered in:  services.AddMediatR(cfg =>

File: source/dotnet/wholesale-api/WebApi.UnitTests/Application/Batches/CreateBatchHandlerTests.cs [hunk 13]
```diff
@@ -0,0 +1,88 @@
...
+    {
+        var period = Periods.January_EuropeCopenhagen_Instant;
+        return new CreateBatchCommand(
+            Contracts.ProcessType.BalanceFixing,
+            new List<string> { "805" },
+            period.PeriodStart.ToDateTimeOffset(),
+            period.PeriodEnd.ToDateTimeOffset());
+    }
+
+    private static Batch CreateBatchFromCommand(CreateBatchCommand command)
```
[the remark this answers is not in the document]
Comment (energinet-ajw) [reply source/dotnet/wholesale-api/WebApi.UnitTests/Application/Batches/CreateBatchHandlerTests.cs:83]: Changed.

File: source/dotnet/wholesale-api/Domain/Domain.csproj [hunk 14]
```diff
@@ -21,6 +21,7 @@ limitations under the License.
   </PropertyGroup>
 
   <ItemGroup>
+    <PackageReference Include="MediatR.Contracts" Version="2.0.1" />
```
[the remark this answers is not in the document]
Comment (energinet-ajw) [reply source/dotnet/wholesale-api/Domain/Domain.csproj:24]: I added the topic to the list of issues we can discuss on Tuesday (11. of April).

File: source/dotnet/wholesale-api/Infrastructure/Persistence/UnitOfWorkPipelineBehavior.cs (hunk 12 above)
Comment (lasrinnil) [reply source/dotnet/wholesale-api/Infrastructure/Persistence/UnitOfWorkPipelineBehavior.cs:30]: All pipeline behaviors are called (in the sequence they were registered in) before the actual handlers are called.

As this behavior is the first one registered - we would call commit after both handlers have been called, and only once.

File: docs/images/create-batch-event-flow.drawio.png
Comment (BjarkeMeier) [inline docs/images/create-batch-event-flow.drawio.png:1]: This diagram is also being added in another PR?

**Pull request closed without merging**